### PR TITLE
Add meta_memcache client

### PR DESCRIPTION
**Pull request by bisho**

### CODE REVIEW

File: src/meta_memcache/base/base_cache_pool.py [hunk 1]
```diff
@@ -0,0 +1,210 @@
...
+        result = c.get_response()
+        if isinstance(result, Value):
+            # TODO: Confirm this works for empty values!
+            data = c.get_value(result.size)
+            if result.size > 0:
+                encoding_id = result.int_flags.get(IntFlag.CLIENT_FLAG, 0)
+                result.value = self._serializer.unserialize(data, encoding_id)
+        return result
+
+    def mg(
```
Comment (ethervoid) [inline src/meta_memcache/base/base_cache_pool.py]: I'll go for better naming here: meta_get, meta_set, meta_delete. If you're not familiar with it is hard to understand

File: src/meta_memcache/base/base_cache_pool.py [hunk 2]
```diff
@@ -0,0 +1,210 @@
...
+        result = self._exec(
+            command=MetaCommand.MG,
+            key=key,
+            flags=flags,
+            int_flags=int_flags,
+            token_flags=token_flags,
+        )
+        if not isinstance(result, (Miss, Value, Success)):
+            raise MemcacheError(
+                f"Unexpected response for Meta Delete command: {result}"
```
Comment (ethervoid) [inline src/meta_memcache/base/base_cache_pool.py]:
```suggestion
                f"Unexpected response for Meta Get command: {result}"
```

File: src/meta_memcache/base/base_cache_pool.py [hunk 3]
```diff
@@ -0,0 +1,210 @@
...
+            command=MetaCommand.MS,
+            key=key,
+            value=encoded_value.data,
+            flags=flags,
+            int_flags=int_flags,
+            token_flags=token_flags,
+        )
+        if not isinstance(result, (Success, NotStored, Conflict, Miss)):
+            raise MemcacheError(
+                f"Unexpected response for Meta Delete command: {result}"
```
Comment (ethervoid) [inline src/meta_memcache/base/base_cache_pool.py]:
```suggestion
                f"Unexpected response for Meta Set command: {result}"
```

File: src/meta_memcache/base/base_cache_pool.py [hunk 4]
```diff
@@ -0,0 +1,210 @@
...
+
+    def mg(
+        self,
+        key: Key,
+        flags: Optional[Set[Flag]] = None,
+        int_flags: Optional[Dict[IntFlag, int]] = None,
+        token_flags: Optional[Dict[TokenFlag, bytes]] = None,
+    ) -> Union[Miss, Value, Success]:
+        result = self._exec(
+            command=MetaCommand.MG,
```
Comment (ethervoid) [inline src/meta_memcache/base/base_cache_pool.py]: Same here `MetaCommand.MetaGet` or `MetaCommand.Get`

File: src/meta_memcache/base/base_cache_pool.py [hunk 5]
```diff
@@ -0,0 +1,210 @@
...
+        self,
+        command: MetaCommand,
+        key: Key,
+        value: Optional[bytes] = None,
+        flags: Optional[Set[Flag]] = None,
+        int_flags: Optional[Dict[IntFlag, int]] = None,
+        token_flags: Optional[Dict[TokenFlag, bytes]] = None,
+    ) -> Union[Miss, Value, Success, NotStored, Conflict]:
+        """
+        Gets a connection for key key and executes the command
```
Comment (ethervoid) [inline src/meta_memcache/base/base_cache_pool.py]:
```suggestion
        Gets a connection for key and executes the command
```

File: src/meta_memcache/base/base_cache_pool.py [hunk 6]
```diff
@@ -0,0 +1,210 @@
...
+        binary_key_encoding_fn: Callable[[Key], bytes],
+        write_failure_tracker: Optional[BaseWriteFailureTracker] = None,
+    ) -> None:
+        self._serializer = serializer
+        self._binary_key_encoding_fn = binary_key_encoding_fn
+        self._write_failure_tracker = write_failure_tracker
+
+    @abstractmethod
+    def _get_pool(self, key: Key) -> ConnectionPool:
+        pass
```
Comment (ethervoid) [inline src/meta_memcache/base/base_cache_pool.py]: What do you think in throwing and `NotImplementedException` here?

File: src/meta_memcache/base/base_cache_pool.py [hunk 7]
```diff
@@ -0,0 +1,210 @@
...
+
+    def ms(
+        self,
+        key: Key,
+        value: Any,  # pyre-ignore[2]
+        ttl: int,
+        flags: Optional[Set[Flag]] = None,
+        int_flags: Optional[Dict[IntFlag, int]] = None,
+        token_flags: Optional[Dict[TokenFlag, bytes]] = None,
+    ) -> Union[Success, NotStored, Conflict, Miss]:
```
Comment (ethervoid) [inline src/meta_memcache/base/base_cache_pool.py:258]: Why don't return just one object with all the data inside like a HTTP response? That why you don't have to do `isinstance` to return a bool and you can return more rich responses if needed

File: src/meta_memcache/base/cache_pool.py [hunk 8]
```diff
@@ -0,0 +1,251 @@
...
+        )
+
+        return isinstance(result, Success)
+
+    def touch(self, key: Key, ttl: int, no_reply: bool = False) -> bool:
+        flags = set()
+        int_flags = {IntFlag.CACHE_TTL: ttl}
+        if no_reply:
+            flags.add(Flag.NOREPLY)
+        result = self.mg(key, flags=flags, int_flags=int_flags)
```
Comment (ethervoid) [inline src/meta_memcache/base/cache_pool.py]: Touch is a meta get? I thought it was a multi set

File: src/meta_memcache/base/cache_pool.py [hunk 9]
```diff
@@ -0,0 +1,251 @@
...
+            IntFlag.MISS_LEASE_TTL: lease_policy.ttl,
+        }
+        if recache_policy:
+            int_flags[IntFlag.RECACHE_TTL] = recache_policy.ttl
+        if touch_ttl is not None and touch_ttl >= 0:
+            int_flags[IntFlag.CACHE_TTL] = touch_ttl
+        i = 0
+        while True:
+            if i > 0:
+                print(
```
Comment (ethervoid) [inline src/meta_memcache/base/cache_pool.py]: Are all these prints for debugging purposes?

**Review by ethervoid — COMMENTED** [review]
Some comments I'll keep reviewing

File: src/meta_memcache/base/base_cache_pool.py (hunk 1 above)
Comment (bisho) [reply src/meta_memcache/base/base_cache_pool.py]: Hum... but they are meant to be internal, just for people really wanting the raw power... If you are strongly opinionated about this I can reconsider...

File: src/meta_memcache/base/base_cache_pool.py (hunk 6 above)
Comment (bisho) [reply src/meta_memcache/base/base_cache_pool.py]: It's abstract, you can't instantiate the class without overriding it. But I can use the ... style that is slightly more common for abstract methods.

File: src/meta_memcache/base/base_cache_pool.py (hunk 7 above)
Comment (bisho) [reply src/meta_memcache/base/base_cache_pool.py:258]: Some responses like miss don't have flags. Only value has size...

In general, it is better for string typing to use this mixed response. This forces callers to do isinstance(Value) before asking `.size` and saves a lot of runtime issues,

File: src/meta_memcache/base/cache_pool.py (hunk 8 above)
Comment (bisho) [reply src/meta_memcache/base/cache_pool.py]: meta commands don't have multiget (you just send several mg without waiting).

And `touch` is implemented with meta get, yes, `mg` can change TTL, and `mg` can not return value, and can also have not reply at all.

File: src/meta_memcache/base/cache_pool.py (hunk 9 above)
Comment (bisho) [reply src/meta_memcache/base/cache_pool.py]: yep, ops

File: src/meta_memcache/base/base_cache_pool.py (hunk 1 above)
Comment (ethervoid) [reply src/meta_memcache/base/base_cache_pool.py]: Yeah I know I'm aiming to improve the readability of the code for people that don't know how it works. I mean calling it mg or meta_get doesn't reduce performance but improves maintainability due to a better readability.

File: src/meta_memcache/base/base_cache_pool.py (hunk 6 above)
Comment (ethervoid) [reply src/meta_memcache/base/base_cache_pool.py]: Oh, I dunno why I thought it was also a class method. nvm

File: src/meta_memcache/base/base_cache_pool.py (hunk 7 above)
Comment (ethervoid) [reply src/meta_memcache/base/base_cache_pool.py:258]: Yeah but I'm usually against having multiple return types that are not inherited or part of an interface. I'll go with something like this:

```python
@dataclass
class MemcachedResponse:
    type: enum
    size: int
    value: Optional[Any]
    flags: Optional[Set[Flag]]

    def __init__(
        self,
        type: enum
        size: Optional[int] = 0,
        value: Optional[Any] = None,  # pyre-ignore[2]
        flags: Optional[Set[Flag]] = None,
        int_flags: Optional[Dict[IntFlag, int]] = None,
        token_flags: Optional[Dict[TokenFlag, bytes]] = None,
    ) -> None:
        super().__init__(flags, int_flags, token_flags)
        self.size = size
        self.value = value
```

and then you can decide what value should have or being set based on the type instead of using `isinstance` and return `-> MemcachedResponse` instead of the Union

File: src/meta_memcache/base/cache_pool.py [hunk 10]
```diff
@@ -0,0 +1,241 @@
...
+
+
+class CachePool(BaseCachePool):
+    def set(
+        self,
+        key: Key,
+        value: Any,  # pyre-ignore[2]
+        ttl: int,
+        no_reply: bool = False,
+        cas: Optional[int] = None,
```
Comment (ethervoid) [inline src/meta_memcache/base/cache_pool.py]: I suppose this is the cas token so I'll rename this a cas_token given that the cas is the operation

File: src/meta_memcache/base/base_cache_pool.py (hunk 1 above)
Comment (bisho) [reply src/meta_memcache/base/base_cache_pool.py]: ack!

File: src/meta_memcache/base/cache_pool.py [hunk 11]
```diff
@@ -0,0 +1,354 @@
...
+        if recache_policy:
+            int_flags[IntFlag.RECACHE_TTL] = recache_policy.ttl
+        if touch_ttl is not None and touch_ttl >= 0:
+            int_flags[IntFlag.CACHE_TTL] = touch_ttl
+
+        result = self.meta_get(key, flags=flags, int_flags=int_flags)
+        if isinstance(result, Value):
+            # It is a hit
+            cas = result.int_flags.get(IntFlag.CAS)
+            if Flag.WIN in result.flags:
```
Comment (ethervoid) [inline src/meta_memcache/base/cache_pool.py:234]: I this accurate given that we don't have lease policy?

File: src/meta_memcache/base/connection_pool.py [hunk 12]
```diff
@@ -0,0 +1,118 @@
...
+        self._errors = 0
+        self._destroyed_counter: itertools.count[int] = itertools.count(start=1)
+        self._destroyed = 0
+        self._marked_down_until: Optional[float] = None
+        self._pool: Queue[MemcacheSocket] = Queue(self._max_pool_size)
+        self._read_buffer_size = read_buffer_size
+        for _ in range(self._initial_pool_size):
+            try:
+                self._pool.put_nowait(self._create_connection())
+            except MemcacheServerError:
```
Comment (ethervoid) [inline src/meta_memcache/base/connection_pool.py:55]: I assume this case would be a connection error and the Queue is going to be empty or not completely created. Is that ok?

File: src/meta_memcache/base/memcache_socket.py [hunk 13]
```diff
@@ -0,0 +1,276 @@
...
+    - The returned memoryview's can be used to unserialize so, for
+      small memcache values, the only allocation is that of serializer
+      building the objects from the bytes in the response.
+    - The buffer can hold several responses, it is only reset
+      when the capacity is less than half, causing data copy of
+      remaining bytes. If there is no remaining bytes (and this is
+      the most likely scenario), the buffer is reset with no cost.
+    """
+
+    def __init__(self, c: socket.socket, buffer_size: int = 4096) -> None:
```
Comment (ethervoid) [inline src/meta_memcache/base/memcache_socket.py]: Can we rename `c` for `conn` or `socket`? It's hard to follow sometimes

**Review by ethervoid — COMMENTED** [review]
More comments, keep reviewing!

File: src/meta_memcache/base/base_cache_pool.py (hunk 7 above)
Comment (bisho) [reply src/meta_memcache/base/base_cache_pool.py:258]: The problem with that is that it is not very type safe.

Returning a `Union`, you force the user of this api to do type checks before using a value, forcing correct error handling.

```
if isinstance(result, Value):
    result.value  # value must exist!!! This is totally type-safe
```

If you really prefer not to have a `Union` I can definitely create an inheritance tree from a common response. But some methods will still use Union as a `get` should not return `stored` for example.

This is also the low-level api, which gives you the raw power of the meta commands and all the data in the responses. And in this layer I still think it is best to use unions for the possible responses. Most users will use the get/set/delete methods that have a much simpler api.

File: src/meta_memcache/base/connection_pool.py (hunk 12 above)
Comment (bisho) [reply src/meta_memcache/base/connection_pool.py:55]: Yeah, the queue is just created in line `50` and `self._initial_pool_size` is `min(initial_pool_size, max_pool_size)`, so it is guaranteed to be `<=` queue max size

File: src/meta_memcache/base/cache_pool.py (hunk 11 above)
Comment (bisho) [reply src/meta_memcache/base/cache_pool.py:234]: Yes, the win/lost flag might be given automatically if something is marked as stale for example, or a lease has been created elsewhere, so you need to handle even when lease policy is not specifically requested in the get command.

File: src/meta_memcache/base/connection_pool.py (hunk 12 above)
Comment (ethervoid) [reply src/meta_memcache/base/connection_pool.py:55]: Yeah but my question is that if an error occurs we're could not reach the `initial_pool_size` so we end up with a Queue with less capacity than the initial value, is that ok?

File: src/meta_memcache/base/base_cache_pool.py (hunk 7 above)
Comment (ethervoid) [reply src/meta_memcache/base/base_cache_pool.py:258]: > The problem with that is that it is not very type safe.

Why?

> If you really prefer not to have a Union I can definitely create an inheritance tree from a common response. But some methods will still use Union as a get should not return stored for example.

I'm not very opinionated, I'm trying to understand why is needed because for me returning different types for the same function is a bad smell and maybe we can have a better approach that is less error prone that having to use `isinstance` when we use that function

File: src/meta_memcache/base/connection_pool.py (hunk 12 above)
Comment (bisho) [reply src/meta_memcache/base/connection_pool.py:55]: Oh, got your point. Yeah, if there are conn errors during initialization, the pool will just start empty.

File: src/meta_memcache/base/base_cache_pool.py (hunk 7 above)
Comment (bisho) [reply src/meta_memcache/base/base_cache_pool.py:258]: Having multiple return types is not an anti pattern. Using unions and isinstance is good over a return object that has many optional values, on with combinations of properties don't make sense, and cause less strict handling on the caller.

In fact, `Optional[Foo]` is a type of union: `Union[None, Foo]`. When you return `Optional[Foo]` you force users to check for noneness before they can use the `Foo` properties (that is, if they use static type checks). This is very powerful to enforce correct error handling. 

Having `Union[Miss,Success,Value]` follows the same principle: forces the caller to do the right handling. Only Value has data inside. With this you can potentially even store a `None` as a serialized value. You can ask callers to check "response type" and use the data only if response type is "value". But you see the problem here, it is not type safe, you can't be sure if users are really checking the reponse type or just using the optional data property directly. With a union, if you try to access `.data` on `Union[Miss, Value]` the type checker will yell at you saying that `data` is not a property of `Miss`.

Furthermore, I was even thinking in creating more response types for when Flag.WIN is returned, again, to make sure wins are handled explicitly, not mishandled as a Value. The win could for example not have data, as you want implementers to threat this as a miss. I didn't do it to mimic more closely the memcached api, and avoid confusion. It is after all an internal layer and I expect users of meta commands to be well aware of the quirks, but like I said, I think Unions on a statically type check python provide huge wins in robustness.

File: src/meta_memcache/base/connection_pool.py (hunk 12 above)
Comment (bisho) [reply src/meta_memcache/base/connection_pool.py:55]: But it will still work properly, mark it down and avoid contant reconnections if keeps being down, and when it recovers, new conns will be added to the pool. It might not reach `initial_pool_size` if the workload needs less connections to operate, but that is not an issue. The initial creation is to reduce latencies when the memcache is first used. If it was down, it is inevitable to have to reconnect at some point.
